doc: llext: improve documentation for the llext subsystem (#75126)
zephyrproject-rtos/zephyr: doc/services/llext/config.rst

Configuration
#############

The following Kconfig options are available for the LLEXT subsystem:

.. _llext_kconf_heap:

Heap size
----------

The LLEXT subsystem requires a heap to be allocated for storing extension
related data. The size of this heap can be configured using the Kconfig option

> :kconfig:option:`CONFIG_LLEXT_HEAP_SIZE`.

Reviewer: maybe make this a "term" just like the ones you have when enumerating the ELF object types in the following section. This way, all Kconfigs will stand out as "top-level"

:kconfig:option:`CONFIG_LLEXT_TYPE_ELF_OBJECT`

        Build and expect relocatable files as binary object type for the LLEXT
        subsystem. A single compiler invocation is used to generate the object
        file.

:kconfig:option:`CONFIG_LLEXT_TYPE_ELF_RELOCATABLE`

        Build and expect relocatable (partially linked) files as the binary
        object type for the LLEXT subsystem. These object files are generated
        by the linker by combining multiple object files into a single one.

:kconfig:option:`CONFIG_LLEXT_TYPE_ELF_SHAREDLIB`

        Build and expect shared libraries as binary object type for the LLEXT
        subsystem. The standard linking process is used to generate the shared
        library from multiple object files.

        This is not currently supported on ARM architectures.

.. _llext_kconf_storage:

Minimize allocations
--------------------

The LLEXT subsystem loading mechanism, by default, uses a seek/read abstraction
and copies all data into allocated memory; this is done to allow the extension
to be loaded from any storage medium. Sometimes, however, data is already in a
buffer in RAM and copying it is not necessary. In this case, the option

> :kconfig:option:`CONFIG_LLEXT_STORAGE_WRITABLE` can be enabled to allow the

Reviewer: ditto re: making this a top-level "term" to stand out more / look more like the ELF Object type section

extension to be loaded directly referencing section data into the buffer. To
be effective, this also requires the use of the :c:struct:`llext_buf_loader` loader.

.. warning::

   The application must ensure that the buffer used to load the extension
   remains allocated until the extension is unloaded.

.. note::

   This will directly modify the contents of the buffer during the link phase.
   Once the extension is unloaded, the buffer must be reloaded before it can be
   used again in a call to :c:func:`llext_load`.

.. note::

   This is currently *required* by the Xtensa architecture.

.. _llext_kconf_slid:

Using SLID to avoid symbol lookups
----------------------------------

When linking an extension to the main Zephyr application, the LLEXT subsystem
uses a series of string comparisons to find the symbols that the extension
needs to link against. This can be optimized by using the Kconfig option

> :kconfig:option:`CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID`.

Reviewer 2: How about something along the lines of:
```
When an extension is loaded, the LLEXT subsystem must find the address of all
the symbols residing in the main application that the extensions {references / imports}.
To this end, the main binary contains a LLEXT-dedicated symbol table, filled with
one symbol-name-to-address mapping entry for each symbol exported by the main
application to {LLEXT / extensions}. This table can then be searched into by the
LLEXT linker during the load of extensions. This process is pretty slow due to
the nature of string comparisons, and the size consumed by the table can become
significant as the number of exported symbols increases.
```

> :kconfig:option:`CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID`.

Reviewer: ditto

When this option is enabled, an extra processing step is performed on the
Zephyr binary and on all extensions: every string in the symbol tables is
converted to a 32-bit hash called `Symbol Link Identifier` (SLID), which is

> stored in the binary. A more efficient hash comparison can then be used to
> quickly find the symbols that the extension needs to link against.

Reviewer 2: ```suggestion
stored in the binary. This speeds up the symbol lookup process by allowing
usage of integer-based comparisons rather than string-based. Another benefit
of SLID-based linking is that storing symbol names in the binary is no longer
necessary, which provides a significant decrease in symbol table size.
```

Author: > Another benefit of SLID-based linking is that storing symbol names in the binary is no longer necessary, which provides a significant decrease in symbol table size.

Absolutely correct, but it's not yet implemented AFAIK. Should we mention this anyway?

Reviewer 2: Actually, this is implemented, but only for Zephyr (because it was trivial).
The trick resides in the linker file:
https://github.com/zephyrproject-rtos/zephyr/blob/d0474cefde27c5fc115a570e388ceec99db53fbd/include/zephyr/linker/llext-sections.ld#L15-L18

Notice that the section is marked as `(COPY)` (AFAIR, this was needed to have the section in ELF - seeing [this](https://sourceware.org/binutils/docs/ld/Output-Section-Type.html), maybe `(NOLOAD)` would be better?), but most importantly, it is not placed in any region. As a result, when the binaries are generated, the section is discarded. [I have empirically verified this myself](https://gist.github.com/mathieuchopstm/e06579bf45bf22b5fd60f1fe3734fd0d) (btw, didn't we have a `llext.simple.readonly_slid_linking` test before?)

This is not implemented in LLEXTs because we can't use the *place all strings in a specific section* trick - it would require manually editing the ELF `.symstr` section instead.

Author: Missed that, thanks for clarifying! Yes, the same trick on the llext side is a lot messier, but also the savings are smaller than in the main binary where it exports _everything_. :+1: 

>  didn't we have a llext.simple.readonly_slid_linking test before?)

It wa axed as part of the rework in #74568, since SLID features should be unrelated to `MMU` and/or `STORAGE_WRITABLE` config options. I kept the "writable" case since that applies to all platforms in one test, but I should have renamed it to `llext.simple.slid_linking` for clarity.

Configuration
#############

The following Kconfig options are available for the LLEXT subsystem:

.. _llext_kconf_heap:

Heap size
----------

The LLEXT subsystem needs a static heap to be allocated for extension related
data. The following option controls this allocation.

:kconfig:option:`CONFIG_LLEXT_HEAP_SIZE`

        Size of the LLEXT heap in kilobytes.

.. note::

   When :ref:`user mode <usermode_api>` is enabled, the heap size must be
   large enough to allow the extension sections to be allocated with the
   alignment required by the architecture.

.. _llext_kconf_type:

ELF object type
---------------

> The LLEXT subsystem supports loading different types of extensions; the type
> can be set by choosing among the following Kconfig options:

Reviewer 2: `grep`'ing the tree, these Kconfig options seem to only affect how extensions are generated and not the LLEXT subsystem. Proposal:
```suggestion
The LLEXT subsystem is able to load various ELF file types as extensions. When
building extensions, which type is built can be selected by choosing among the
following Kconfig options:
```

(out-of-scope, but shouldn't this choice be turned into a CMake function argument rather than Kconfig, for this reason?)

Author: > these Kconfig options seem to only affect how extensions are generated and not the LLEXT subsystem

This is currently true, however that might be due to the minimal overlap between targets and ELF types. I think properly supporting dynamic loading might benefit from knowing the ELF type in advance, to avoid littering address-related code with "is this a VMA or file offset? check the elf type". That's why I preferred to be a little bit more restrictive than necessary.

Reviewer 3: That seems entirely reasonable to me! Agreed

to be loaded from any storage medium. Sometimes, however, data is already in a
buffer in RAM and copying it is not necessary. The following option allows the
LLEXT subsystem to optimize memory footprint in this case.

:kconfig:option:`CONFIG_LLEXT_STORAGE_WRITABLE`

        Allow the extension to be loaded by directly referencing section data
        into the ELF buffer. To be effective, this requires the use of an ELF
        loader that supports the ``peek`` functionality, such as the
        :c:struct:`llext_buf_loader`.

        .. warning::

           The application must ensure that the buffer used to load the
           extension remains allocated until the extension is unloaded.

        .. note::

           This will directly modify the contents of the buffer during the link
           phase. Once the extension is unloaded, the buffer must be reloaded
           before it can be used again in a call to :c:func:`llext_load`.

        .. note::

           This is currently *required* by the Xtensa architecture.

.. _llext_kconf_slid:

Using SLID for symbol lookups
-----------------------------

When an extension is loaded, the LLEXT subsystem must find the address of all

> the symbols residing in the main application that the extensions references.

Reviewer 2: Typo:
```suggestion
the symbols residing in the main application that the extension references.
```

with one symbol-name-to-address mapping entry for each symbol exported by the
main application to extensions. This table can then be searched into by the
LLEXT linker at extension load time. This process is pretty slow due to the
nature of string comparisons, and the size consumed by the table can become
significant as the number of exported symbols increases.

:kconfig:option:`CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID`

        Perform an extra processing step on the Zephyr binary and on all
        extensions being built, converting every string in the symbol tables to
        a pointer-sized hash called `Symbol Link Identifier` (SLID), which is
        stored in the binary.

        This speeds up the symbol lookup process by allowing usage of
        integer-based comparisons rather than string-based ones. Another
        benefit of SLID-based linking is that storing symbol names in the
        binary is no longer necessary, which provides a significant decrease in
        symbol table size.

        .. note::

           This option is not currently compatible with the :ref:`LLEXT EDK
           <llext_build_edk>`.

> 

Reviewer 2: Maybe an additional sidenote could be worth adding:
```suggestion
        .. note::

           This option must have the same value when building the kernel and all
           extensions to be loaded by said kernel. For example, if the kernel
           is built with CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID=y, it is forbidden
           to load an extension built with CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID=n.
```

(ATM, the converse case (kernel=n, extension=y) works, but this may no longer apply in the future if we strip extensions as well)

        Size of the LLEXT heap in kilobytes.

.. note::

   When :ref:`user mode <usermode_api>` is enabled, the heap size must be
   large enough to allow the extension sections to be allocated with the
   alignment required by the architecture.

.. _llext_kconf_type:

ELF object type
---------------

The LLEXT subsystem supports loading different types of extensions; the type
can be set by choosing among the following Kconfig options:

:kconfig:option:`CONFIG_LLEXT_TYPE_ELF_OBJECT`

        Build and expect relocatable files as binary object type for the LLEXT
        subsystem. A single compiler invocation is used to generate the object
        file.

:kconfig:option:`CONFIG_LLEXT_TYPE_ELF_RELOCATABLE`

        Build and expect relocatable (partially linked) files as the binary
        object type for the LLEXT subsystem. These object files are generated
        by the linker by combining multiple object files into a single one.

:kconfig:option:`CONFIG_LLEXT_TYPE_ELF_SHAREDLIB`

        Build and expect shared libraries as binary object type for the LLEXT
        subsystem. The standard linking process is used to generate the shared
        library from multiple object files.

>         This is not currently supported on ARM architectures.

Reviewer 3: I'd make this a ```.. note::```

> .. _llext_kconf_storage:

Reviewer 3: add the two letters to full spell kconfig for the labels here would be just fine by me

Minimize allocations
--------------------

The LLEXT subsystem loading mechanism, by default, uses a seek/read abstraction
and copies all data into allocated memory; this is done to allow the extension
to be loaded from any storage medium. Sometimes, however, data is already in a
buffer in RAM and copying it is not necessary. The following option allows the
LLEXT subsystem to optimize memory footprint in this case.

:kconfig:option:`CONFIG_LLEXT_STORAGE_WRITABLE`

        Allow the extension to be loaded by directly referencing section data
        into the ELF buffer. To be effective, this requires the use of an ELF
        loader that supports the ``peek`` functionality, such as the
        :c:struct:`llext_buf_loader`.

        .. warning::

           The application must ensure that the buffer used to load the
           extension remains allocated until the extension is unloaded.

        .. note::

           This will directly modify the contents of the buffer during the link
           phase. Once the extension is unloaded, the buffer must be reloaded
           before it can be used again in a call to :c:func:`llext_load`.

        .. note::

>            This is currently *required* by the Xtensa architecture.

Reviewer 3: If this is true we should file a bug/feature issue imo if we haven't already for this and maybe reference it here
